tunnel/intra/doh/doh.go

```diff
@@ -190,34 +196,101 @@ func (e *queryError) Unwrap() error {
 // Given a raw DNS query (including the query ID), this function sends the
 // query.  If the query is successful, it returns the response and a nil qerr.  Otherwise,
 // it returns a nil response and a qerr with a status value indicating the cause.
-// Independent of the query's success or failure, this function also returns the IP
-// address of the server on a best-effort basis, returning the empty string if the address
-// could not be determined.
-func (t *transport) doQuery(q []byte) (response []byte, server string, qerr error) {
+// Independent of the query's success or failure, this function also returns the
+// address of the server on a best-effort basis, or nil if the address could not
+// be determined.
+func (t *transport) doQuery(q []byte) (response []byte, server *net.TCPAddr, qerr error) {
 	if len(q) < 2 {
 		qerr = &queryError{BadQuery, fmt.Errorf("Query length is %d", len(q))}
 		return
 	}
 	id0, id1 := q[0], q[1]
+	id := int(id0)*256 + int(id1)
 	// Zero out the query ID.
 	q[0], q[1] = 0, 0
-	req, err := http.NewRequest("POST", t.url, bytes.NewBuffer(q))
+	req, err := http.NewRequest(http.MethodPost, t.url, bytes.NewBuffer(q))
 	if err != nil {
 		qerr = &queryError{InternalError, err}
 		return
 	}
 
+	hostname := t.hostname
+
+	// The connection used for this request.  If the request fails, we will close
+	// this socket, in case it is no longer functioning.
+	var conn net.Conn
+
+	// Error cleanup function.  If the query fails, this function will close the
+	// underlying socket and disconfirm the server IP.  Empirically, sockets often
+	// become unresponsive after a network change, causing timeouts on all requests.
+	defer func() {
+		if qerr == nil {
+			return
+		}
+		log.Infof("%d Query failed: %v", id, qerr)
+		if server != nil {
+			log.Debugf("%d Disconfirming %s", id, server.IP.String())
+			t.ips.Get(hostname).Disconfirm(server.IP)
```

Comment:
In the event of a timeout due to a network change, we would reconfirm the server IP when the query is reissued (and successful), right?

Reply:
Yes, at the end of `doQuery`.